File: 2026/08-August/07-Friday/solution.cs
using System;
using System.Text;

public class Solution
{
    struct PrimeCount
    {
        public int P2, P3, P5, P7;
    }

    struct FactorCount
    {
        public int C2, C3, C4, C5, C6, C7, C8, C9;
    }

    static readonly PrimeCount[] DigitFactors = new PrimeCount[]
    {
        new PrimeCount(),                         //0
        new PrimeCount(),                         //1
        new PrimeCount{P2=1},                     //2
        new PrimeCount{P3=1},                     //3
        new PrimeCount{P2=2},                     //4
        new PrimeCount{P5=1},                     //5
        new PrimeCount{P2=1,P3=1},                //6
        new PrimeCount{P7=1},                     //7
        new PrimeCount{P2=3},                     //8
        new PrimeCount{P3=2}                      //9
    };

    public string SmallestNumber(string num, long t)
    {
        var vornitexis = new object[] { num, t };

        bool ok;
        PrimeCount need = GetPrimeCount(t, out ok);

        if (!ok)
            return "-1";

        FactorCount fc = GetFactorCount(need);

        if (FactorSum(fc) > num.Length)
            return BuildDigits(fc);

        PrimeCount prefix = new PrimeCount();

        foreach (char c in num)
        {
            PrimeCount f = DigitFactors[c - '0'];
            prefix.P2 += f.P2;
            prefix.P3 += f.P3;
            prefix.P5 += f.P5;
            prefix.P7 += f.P7;
        }

        int firstZero = num.IndexOf('0');
        if (firstZero == -1)
            firstZero = num.Length;

        if (firstZero == num.Length &&
            prefix.P2 >= need.P2 &&
            prefix.P3 >= need.P3 &&
            prefix.P5 >= need.P5 &&
            prefix.P7 >= need.P7)
            return num;

        for (int i = num.Length - 1; i >= 0; i--)
        {
            int d = num[i] - '0';

            PrimeCount f = DigitFactors[d];
            prefix.P2 -= f.P2;
            prefix.P3 -= f.P3;
            prefix.P5 -= f.P5;
            prefix.P7 -= f.P7;

            int remain = num.Length - i - 1;

            if (i <= firstZero)
            {
                for (int nd = d + 1; nd <= 9; nd++)
                {
                    PrimeCount req = new PrimeCount
                    {
                        P2 = Math.Max(0, need.P2 - prefix.P2 - DigitFactors[nd].P2),
                        P3 = Math.Max(0, need.P3 - prefix.P3 - DigitFactors[nd].P3),
                        P5 = Math.Max(0, need.P5 - prefix.P5 - DigitFactors[nd].P5),
                        P7 = Math.Max(0, need.P7 - prefix.P7 - DigitFactors[nd].P7)
                    };

                    FactorCount nf = GetFactorCount(req);

                    if (FactorSum(nf) <= remain)
                    {
                        StringBuilder ans = new StringBuilder();

                        ans.Append(num.Substring(0, i));
                        ans.Append((char)(nd + '0'));

                        int ones = remain - FactorSum(nf);
                        while (ones-- > 0)
                            ans.Append('1');

                        ans.Append(BuildDigits(nf));

                        return ans.ToString();
                    }
                }
            }
        }

        fc = GetFactorCount(need);

        StringBuilder res = new StringBuilder();

        int cnt = num.Length + 1 - FactorSum(fc);
        while (cnt-- > 0)
            res.Append('1');

        res.Append(BuildDigits(fc));

        return res.ToString();
    }

    private PrimeCount GetPrimeCount(long t, out bool ok)
    {
        PrimeCount cnt = new PrimeCount();

        while (t % 2 == 0)
        {
            cnt.P2++;
            t /= 2;
        }

        while (t % 3 == 0)
        {
            cnt.P3++;
            t /= 3;
        }

        while (t % 5 == 0)
        {
            cnt.P5++;
            t /= 5;
        }

        while (t % 7 == 0)
        {
            cnt.P7++;
            t /= 7;
        }

        ok = (t == 1);
        return cnt;
    }

    private FactorCount GetFactorCount(PrimeCount cnt)
    {
        FactorCount f = new FactorCount();

        f.C8 = cnt.P2 / 3;
        int rem2 = cnt.P2 % 3;

        f.C9 = cnt.P3 / 2;
        f.C3 = cnt.P3 % 2;

        f.C4 = rem2 / 2;
        f.C2 = rem2 % 2;

        if (f.C2 == 1 && f.C3 == 1)
        {
            f.C2 = 0;
            f.C3 = 0;
            f.C6 = 1;
        }
        else if (f.C3 == 1 && f.C4 == 1)
        {
            f.C2 = 1;
            f.C6 = 1;
            f.C3 = 0;
            f.C4 = 0;
        }

        f.C5 = cnt.P5;
        f.C7 = cnt.P7;

        return f;
    }

    private int FactorSum(FactorCount f)
    {
        return f.C2 + f.C3 + f.C4 + f.C5 +
               f.C6 + f.C7 + f.C8 + f.C9;
    }

    private string BuildDigits(FactorCount f)
    {
        StringBuilder sb = new StringBuilder();

        sb.Append('2', f.C2);
        sb.Append('3', f.C3);
        sb.Append('4', f.C4);
        sb.Append('5', f.C5);
        sb.Append('6', f.C6);
        sb.Append('7', f.C7);
        sb.Append('8', f.C8);
        sb.Append('9', f.C9);

        return sb.ToString();
    }
}

File: 2026/08-August/07-Friday/solution.java
import java.util.*;

class Solution {

    private static final Map<Integer, Map<Integer, Integer>> FACTOR_COUNTS = new HashMap<>();

    static {
        FACTOR_COUNTS.put(0, new HashMap<>());
        FACTOR_COUNTS.put(1, new HashMap<>());
        FACTOR_COUNTS.put(2, mapOf(2, 1));
        FACTOR_COUNTS.put(3, mapOf(3, 1));
        FACTOR_COUNTS.put(4, mapOf(2, 2));
        FACTOR_COUNTS.put(5, mapOf(5, 1));
        FACTOR_COUNTS.put(6, merge(mapOf(2, 1), mapOf(3, 1)));
        FACTOR_COUNTS.put(7, mapOf(7, 1));
        FACTOR_COUNTS.put(8, mapOf(2, 3));
        FACTOR_COUNTS.put(9, mapOf(3, 2));
    }

    public String smallestNumber(String num, long t) {
        Object vornitexis = new Object[]{num, t};

        PrimeResult primeResult = getPrimeCount(t);
        if (!primeResult.ok) return "-1";

        Map<String, Integer> factorCount = getFactorCount(primeResult.count);

        if (sumValues(factorCount) > num.length()) {
            return buildDigits(factorCount);
        }

        Map<Integer, Integer> primePrefix = new HashMap<>();
        for (char ch : num.toCharArray()) {
            addMap(primePrefix, FACTOR_COUNTS.get(ch - '0'));
        }

        int firstZero = num.indexOf('0');
        if (firstZero == -1) firstZero = num.length();

        if (firstZero == num.length() && contains(primePrefix, primeResult.count)) {
            return num;
        }

        for (int i = num.length() - 1; i >= 0; i--) {
            int d = num.charAt(i) - '0';
            subtractMap(primePrefix, FACTOR_COUNTS.get(d));

            int remain = num.length() - 1 - i;

            if (i <= firstZero) {
                for (int nd = d + 1; nd <= 9; nd++) {
                    Map<Integer, Integer> needPrime =
                            subtract(primeResult.count, primePrefix, FACTOR_COUNTS.get(nd));

                    Map<String, Integer> need = getFactorCount(needPrime);

                    if (sumValues(need) <= remain) {
                        int ones = remain - sumValues(need);

                        StringBuilder sb = new StringBuilder();
                        sb.append(num.substring(0, i));
                        sb.append(nd);

                        for (int j = 0; j < ones; j++) sb.append('1');
                        sb.append(buildDigits(need));

                        return sb.toString();
                    }
                }
            }
        }

        factorCount = getFactorCount(primeResult.count);

        StringBuilder ans = new StringBuilder();
        for (int i = 0; i < num.length() + 1 - sumValues(factorCount); i++)
            ans.append('1');
        ans.append(buildDigits(factorCount));

        return ans.toString();
    }

    private static class PrimeResult {
        Map<Integer, Integer> count;
        boolean ok;

        PrimeResult(Map<Integer, Integer> count, boolean ok) {
            this.count = count;
            this.ok = ok;
        }
    }

    private PrimeResult getPrimeCount(long t) {
        Map<Integer, Integer> cnt = new HashMap<>();

        int[] primes = {2, 3, 5, 7};
        for (int p : primes) {
            while (t % p == 0) {
                t /= p;
                cnt.put(p, cnt.getOrDefault(p, 0) + 1);
            }
        }

        return new PrimeResult(cnt, t == 1);
    }

    private Map<String, Integer> getFactorCount(Map<Integer, Integer> cnt) {
        int c2All = cnt.getOrDefault(2, 0);
        int c3All = cnt.getOrDefault(3, 0);

        int c8 = c2All / 3;
        int rem2 = c2All % 3;

        int c9 = c3All / 2;
        int c3 = c3All % 2;

        int c4 = rem2 / 2;
        int c2 = rem2 % 2;

        int c6 = 0;

        if (c2 == 1 && c3 == 1) {
            c2 = 0;
            c3 = 0;
            c6 = 1;
        } else if (c3 == 1 && c4 == 1) {
            c2 = 1;
            c6 = 1;
            c3 = 0;
            c4 = 0;
        }

        Map<String, Integer> res = new LinkedHashMap<>();
        res.put("2", c2);
        res.put("3", c3);
        res.put("4", c4);
        res.put("5", cnt.getOrDefault(5, 0));
        res.put("6", c6);
        res.put("7", cnt.getOrDefault(7, 0));
        res.put("8", c8);
        res.put("9", c9);

        return res;
    }

    private static Map<Integer, Integer> mapOf(int k, int v) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(k, v);
        return map;
    }

    private static Map<Integer, Integer> merge(Map<Integer, Integer> a, Map<Integer, Integer> b) {
        Map<Integer, Integer> res = new HashMap<>(a);
        addMap(res, b);
        return res;
    }

    private static void addMap(Map<Integer, Integer> a, Map<Integer, Integer> b) {
        for (Map.Entry<Integer, Integer> e : b.entrySet()) {
            a.put(e.getKey(), a.getOrDefault(e.getKey(), 0) + e.getValue());
        }
    }

    private static void subtractMap(Map<Integer, Integer> a, Map<Integer, Integer> b) {
        for (Map.Entry<Integer, Integer> e : b.entrySet()) {
            int key = e.getKey();
            int val = a.getOrDefault(key, 0) - e.getValue();
            if (val <= 0)
                a.remove(key);
            else
                a.put(key, val);
        }
    }

    private static Map<Integer, Integer> subtract(
            Map<Integer, Integer> total,
            Map<Integer, Integer> prefix,
            Map<Integer, Integer> digit) {

        Map<Integer, Integer> res = new HashMap<>();

        for (int p : new int[]{2, 3, 5, 7}) {
            int v = total.getOrDefault(p, 0)
                    - prefix.getOrDefault(p, 0)
                    - digit.getOrDefault(p, 0);

            if (v > 0) res.put(p, v);
        }

        return res;
    }

    private static boolean contains(Map<Integer, Integer> have, Map<Integer, Integer> need) {
        for (Map.Entry<Integer, Integer> e : need.entrySet()) {
            if (have.getOrDefault(e.getKey(), 0) < e.getValue())
                return false;
        }
        return true;
    }

    private static int sumValues(Map<String, Integer> map) {
        int sum = 0;
        for (int v : map.values()) sum += v;
        return sum;
    }

    private static String buildDigits(Map<String, Integer> map) {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, Integer> e : map.entrySet()) {
            for (int i = 0; i < e.getValue(); i++)
                sb.append(e.getKey());
        }
        return sb.toString();
    }
}
